net/libfab/libfab.c

```diff
@@ -0,0 +1,2829 @@
+	cnt = libfab_check_for_comp(cq, buf, NULL, NULL);
+	for (i = 0; i < cnt; i++) {
+		if (buf[i] != NULL) {
+			aep = libfab_aep_get(buf[i]->fb_txctx);
+			if (M0_IN(buf[i]->fb_nb->nb_qtype,
+				  (M0_NET_QT_ACTIVE_BULK_RECV,
+				   M0_NET_QT_ACTIVE_BULK_SEND)))
+				--aep->aep_bulk_cnt;
+			buf[i]->fb_wr_comp_cnt++;
+			if (buf[i]->fb_wr_comp_cnt >= buf[i]->fb_wr_cnt) {
+				libfab_target_notify(buf[i], aep);
+				libfab_buf_done(buf[i], 0);
+			}
+		}
+	}
+}
+
+/**
+ * Used to poll for connection events, completion events and process the queued
+ * bulk buffer operations.
+ */
+static void libfab_poller(struct m0_fab__tm *tm)
+{
+	struct m0_net_end_point  *net;
+	struct m0_fab__ep        *xep;
+	struct m0_fab__active_ep *aep;
+	struct fid_cq            *cq;
+	struct epoll_event        ev;
+	int                       ev_cnt;
+
+	while (tm->ftm_shutdown == false) {
+		usleep(0);
+		ev_cnt = epoll_wait(tm->ftm_epfd, &ev, 1, FAB_WAIT_FD_TMOUT);
+
+		while(1) {
+			m0_mutex_lock(&tm->ftm_endlock);
+			if (tm->ftm_shutdown)
+				break;
+			else if (m0_mutex_trylock(&tm->ftm_ntm->ntm_mutex) 
+									 != 0) {
+				m0_mutex_unlock(&tm->ftm_endlock);
+			} else
+				break;
+		}
+		
+		m0_mutex_unlock(&tm->ftm_endlock);
+		
+		if (tm->ftm_shutdown)
+			break;
+		
+		M0_ASSERT(libfab_tm_is_locked(tm) && libfab_tm_invariant(tm));
+
+		if (ev_cnt > 0) {
+			libfab_handle_connect_request_events(tm);
+			libfab_txep_comp_read(tm->ftm_tx_cq);
+
+			m0_tl_for(m0_nep, &tm->ftm_ntm->ntm_end_points, net) {
+				xep = libfab_ep_net(net);
+				aep = libfab_aep_get(xep);
+				if (aep != NULL) {
+					libfab_txep_event_check(xep, aep, tm);
+					cq = aep->aep_rx_res.frr_cq;
+					libfab_rxep_comp_read(cq, xep);
+				}
+			} m0_tl_endfor;
+		}
+
+		libfab_bulk_buf_process(tm);
+		libfab_tm_buf_timeout(tm);
+		libfab_tm_buf_done(tm);
+
+		M0_ASSERT(libfab_tm_invariant(tm));
+		libfab_tm_unlock(tm);
+	}
+}
+
+/** 
+ * Converts network end-point to its libfabric structure.
+ */
+static struct m0_fab__ep *libfab_ep_net(struct m0_net_end_point *net)
+{
+	return container_of(net, struct m0_fab__ep, fep_nep);
+}
+
+/**
+ * Compares the endpoint name with the passed name string and
+ * returns true if equal, or else returns false
+ * If name is null then the ipaddress and port fields are compared to check for
+ * a matching endpoint.
+ */
+static bool libfab_ep_cmp(struct m0_fab__ep *ep, const char *name,
+			  struct m0_fab__ep_name *epname)
+{
+	bool ret = false;
+	if (name == NULL) {
+		if ((strcmp(ep->fep_name.fen_addr, epname->fen_addr) == 0) &&
+		    (strcmp(ep->fep_name.fen_port, epname->fen_port) == 0))
+			ret = true;
+	}
+	else {
+		if (strcmp(ep->fep_name.fen_str_addr, name) == 0)
+			ret = true;
+	}
+	return ret;
+}
+
+/**
+ * Search for the ep in the existing ep list using one of the following -
+ *   1) Name in str format      OR
+ *   2) ipaddr and port 
+ * If found then return the ep structure, or else create a new endpoint
+ * with the name
```

Comment:
Please add comment about the return value.

Reply:
Updated